README.md
LIBLOHS
=======

LIBLOHS is a library (driver) to control LOHS Readers (for now there is support only to 3x models... more in the future). Now it's under development and it's only for developers. (I hope a fully functional release very soon).

Changelog
---------

0.01:
        - Initial release/freshmeat release
        - Developers only release

0.02:
        - Code completly rewriten.

0.03:
        - Added timeval struct support in lohs.c:strRX()
        - Bug Fixed: lohs.c:strTX missing strncpy()


I'm without a LOHS scanner right now, if you want some kind of development, please borrow a reader to me. 

Thanks to
---------

Snape, Stormbringer, Andre, dm_, rct from hsinformatica


Changelog
0.01:
	- Initial release/freshmeat release
	- Developers only release
0.02:
	- Code completly rewriten.
0.03:
	- Added timeval struct support in lohs.c:strRX()
	- Bug Fixed: lohs.c:strTX missing strncpy()



lohs.h
/* PROGRAM:	liblohs.so
 * PURPOSE:	Specify all functions in lohs protocol
 * AUTHOR:	Igor Morgado <igor@void.com.br>
 * DATE:	2003-10-13
 * REVISED:	$Log: lohs.h,v $
 * REVISED:	Revision 1.1.1.1  2003/10/14 18:37:21  imorgado
 * REVISED:	Initial import.
 * REVISED:	
 * Copyright 2003
 * Redistributable under the terms of the GNU General Public Licence (GPL)
 */

extern int open_port(const char *port);
extern int close_port(int fd);
extern int set_serialline(int fd,
		const char *bauds, 
		const char *flowcontrol, 
		const char *databits,
		const char *stopbits);
extern int reset_serialline(int fd);

extern int LOHS_cancel_wait(int fd);
extern int LOHS_set_default(int fd);
extern int LOHS_reset(int fd);
extern int LOHS_start_read(int fd);
extern int LOHS_buffer_tx(int fd, int bnumber);
extern int LOHS_beep(int fd);
extern int LOHS_com_test(int fd, int value);
extern int LOHS_read_times(int fd, int ntimes);
extern int LOHS_verify_type(int fd, int vtype);
extern int LOHS_clocks_number(int fd, int nclock);
extern int LOHS_read_mask(int fd, char *mask);
extern int LOHS_flow_control(int fd, int fcontrol);
extern int LOHS_status(int fd);
extern char * LOHS_text_block(int fd);
extern int LOHS_display_text(int fd, const char *mesg);
extern int LOHS_show_version(int fd);
extern int LOHS_version_info(int fd);



hardtry.c
/* Para testes de conexao */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <termios.h>
#include <sys/select.h>


#include "global.h"

struct termios oldtio, newtio;

fd_set fds;
struct timeval tv;
int retval;

char mesg[4], blocked_mesg[7], r_buf[1];
uint16 blk;

int fd, wout, r;

/* Calculate 16 bit CRC checksum on specified block */
                                                                                                                             
#define TBLSIZE         256
#define CRC16_INIT      0

uint16 CRC16Table[TBLSIZE] = {
	0x0000, 0xc0c1, 0xc181, 0x0140, 0xc301, 0x03c0, 0x0280, 0xc241,
	0xc601, 0x06c0, 0x0780, 0xc741, 0x0500, 0xc5c1, 0xc481, 0x0440,
	0xcc01, 0x0cc0, 0x0d80, 0xcd41, 0x0f00, 0xcfc1, 0xce81, 0x0e40,
	0x0a00, 0xcac1, 0xcb81, 0x0b40, 0xc901, 0x09c0, 0x0880, 0xc841,
	0xd801, 0x18c0, 0x1980, 0xd941, 0x1b00, 0xdbc1, 0xda81, 0x1a40,
	0x1e00, 0xdec1, 0xdf81, 0x1f40, 0xdd01, 0x1dc0, 0x1c80, 0xdc41,
	0x1400, 0xd4c1, 0xd581, 0x1540, 0xd701, 0x17c0, 0x1680, 0xd641,
	0xd201, 0x12c0, 0x1380, 0xd341, 0x1100, 0xd1c1, 0xd081, 0x1040,
	0xf001, 0x30c0, 0x3180, 0xf141, 0x3300, 0xf3c1, 0xf281, 0x3240,
	0x3600, 0xf6c1, 0xf781, 0x3740, 0xf501, 0x35c0, 0x3480, 0xf441,
	0x3c00, 0xfcc1, 0xfd81, 0x3d40, 0xff01, 0x3fc0, 0x3e80, 0xfe41,
	0xfa01, 0x3ac0, 0x3b80, 0xfb41, 0x3900, 0xf9c1, 0xf881, 0x3840,
	0x2800, 0xe8c1, 0xe981, 0x2940, 0xeb01, 0x2bc0, 0x2a80, 0xea41,
	0xee01, 0x2ec0, 0x2f80, 0xef41, 0x2d00, 0xedc1, 0xec81, 0x2c40,
	0xe401, 0x24c0, 0x2580, 0xe541, 0x2700, 0xe7c1, 0xe681, 0x2640,
	0x2200, 0xe2c1, 0xe381, 0x2340, 0xe101, 0x21c0, 0x2080, 0xe041,
	0xa001, 0x60c0, 0x6180, 0xa141, 0x6300, 0xa3c1, 0xa281, 0x6240,
	0x6600, 0xa6c1, 0xa781, 0x6740, 0xa501, 0x65c0, 0x6480, 0xa441,
	0x6c00, 0xacc1, 0xad81, 0x6d40, 0xaf01, 0x6fc0, 0x6e80, 0xae41,
	0xaa01, 0x6ac0, 0x6b80, 0xab41, 0x6900, 0xa9c1, 0xa881, 0x6840,
	0x7800, 0xb8c1, 0xb981, 0x7940, 0xbb01, 0x7bc0, 0x7a80, 0xba41,
	0xbe01, 0x7ec0, 0x7f80, 0xbf41, 0x7d00, 0xbdc1, 0xbc81, 0x7c40,
	0xb401, 0x74c0, 0x7580, 0xb541, 0x7700, 0xb7c1, 0xb681, 0x7640,
	0x7200, 0xb2c1, 0xb381, 0x7340, 0xb101, 0x71c0, 0x7080, 0xb041,
	0x5000, 0x90c1, 0x9181, 0x5140, 0x9301, 0x53c0, 0x5280, 0x9241,
	0x9601, 0x56c0, 0x5780, 0x9741, 0x5500, 0x95c1, 0x9481, 0x5440,
	0x9c01, 0x5cc0, 0x5d80, 0x9d41, 0x5f00, 0x9fc1, 0x9e81, 0x5e40,
	0x5a00, 0x9ac1, 0x9b81, 0x5b40, 0x9901, 0x59c0, 0x5880, 0x9841,
	0x8801, 0x48c0, 0x4980, 0x8941, 0x4b00, 0x8bc1, 0x8a81, 0x4a40,
	0x4e00, 0x8ec1, 0x8f81, 0x4f40, 0x8d01, 0x4dc0, 0x4c80, 0x8c41,
	0x4400, 0x84c1, 0x8581, 0x4540, 0x8701, 0x47c0, 0x4680, 0x8641,
	0x8201, 0x42c0, 0x4380, 0x8341, 0x4100, 0x81c1, 0x8081, 0x4040
};

uint16
BlockCRC16(char *block, int16 len) 
{
	uint16 crc, i; 

	crc=CRC16_INIT;

	for (i=0; i<len; i++)
	{
		crc=(int16)((crc>>8)^(uint16)CRC16Table[(crc^(uint16)block[i])&0x00ff]);
	}

	return (crc);
}


int main(void) {

	/* We will open the device */
	fd = open ("/dev/ttyS0" , O_RDWR | O_NOCTTY);

	if(errno != 0)
		perror ("open port: Unable to open device"), exit(EXIT_FAILURE);
	else fcntl (fd, F_SETFL, 0);
	
	/* Set serial line */
	tcgetattr (fd, &oldtio);
        memset(&newtio, '0', sizeof(newtio));

	newtio.c_cflag = B9600 | CRTSCTS | CS8 | CSTOPB | CLOCAL | CREAD;
	newtio.c_iflag = oldtio.c_iflag;
	newtio.c_oflag = oldtio.c_oflag;
	newtio.c_lflag = oldtio.c_lflag;

	/* Clean the line and activate settings */
	tcsetattr(fd, TCSANOW, &newtio);
	tcflush(fd, TCIFLUSH);

	/* With fd opened and serial ok.
	 * We will try to send a beep
	 */

	strncpy (mesg, "\x02\x06\x03\x03", 4); 
	/*strncpy(mesg, "ATDT99215952", 4);*/
	strcpy (blocked_mesg, mesg);

	printf("Mensagem original: %s\n", mesg);
	
	/* We have to add CRC16 block to the message */
	blk = BlockCRC16((char *)mesg, 4);
	blocked_mesg[4]=(byte) (blk >> 8) & 0xff;
	blocked_mesg[5]=(byte)  blk       & 0xff;
	blocked_mesg[6]='\0';
	printf("Mensagem blocada: %s\n", blocked_mesg);
	
	printf("Enviando comando: %s de %i bytes\n", blocked_mesg, strlen(blocked_mesg));
	wout = write (fd, blocked_mesg, 7);
	
	if (wout < 0)
		perror ("Error: Writing: BEEP CMD\n");
	

	printf("Lendo ACK\n");

	FD_ZERO(&fds);
	FD_SET(fd, &fds);

	tv.tv_sec = 5;
	tv.tv_usec = 0;

	retval = select(fd+1, &fds, NULL, NULL, &tv);

	if (retval == -1)
		perror("reading: select()");
	else if (retval)
		r=read (fd, r_buf, 1); 
	else
		printf(" Cannot read in 5 seconds from device\n"); 

	return 0;
	
	/* Clean everything */
	tcsetattr(fd, TCSANOW, &oldtio);
	close(fd);

	return 0;
}


crc16-ansi.c
/* PROGRAM:     liblohs.so
 * PURPOSE:     Functions to calculate CRC checksum
 *		using ANSI Polynom - X^16+X^15+X^2+1
 * AUTHOR:      Igor Morgado <igor@void.com.br>
 * DATE:        2003/06/03
 *
 * REVISED:     $Log: crc16-ansi.c,v $
 * REVISED:     Revision 1.1.1.1  2003/10/14 18:37:20  imorgado
 * REVISED:     Initial import.
 * REVISED:
 * REVISED:     Revision 1.0  2003/06/03 17:23 BRT
 * REVISED:     Initial revision
 * REVISED:
 * Copyright 2003 - Igor Morgado
 * Redistributable under the terms of the GNU General Public Licence (GPL)
 */

#include "global.h"

/* Calculate 16 bit CRC checksum on specified block */
                                                                                                                             
#define TBLSIZE         256
#define CRC16_INIT      0

uint16 CRC16Table[TBLSIZE] = {
	0x0000, 0xc0c1, 0xc181, 0x0140, 0xc301, 0x03c0, 0x0280, 0xc241,
	0xc601, 0x06c0, 0x0780, 0xc741, 0x0500, 0xc5c1, 0xc481, 0x0440,
	0xcc01, 0x0cc0, 0x0d80, 0xcd41, 0x0f00, 0xcfc1, 0xce81, 0x0e40,
	0x0a00, 0xcac1, 0xcb81, 0x0b40, 0xc901, 0x09c0, 0x0880, 0xc841,
	0xd801, 0x18c0, 0x1980, 0xd941, 0x1b00, 0xdbc1, 0xda81, 0x1a40,
	0x1e00, 0xdec1, 0xdf81, 0x1f40, 0xdd01, 0x1dc0, 0x1c80, 0xdc41,
	0x1400, 0xd4c1, 0xd581, 0x1540, 0xd701, 0x17c0, 0x1680, 0xd641,
	0xd201, 0x12c0, 0x1380, 0xd341, 0x1100, 0xd1c1, 0xd081, 0x1040,
	0xf001, 0x30c0, 0x3180, 0xf141, 0x3300, 0xf3c1, 0xf281, 0x3240,
	0x3600, 0xf6c1, 0xf781, 0x3740, 0xf501, 0x35c0, 0x3480, 0xf441,
	0x3c00, 0xfcc1, 0xfd81, 0x3d40, 0xff01, 0x3fc0, 0x3e80, 0xfe41,
	0xfa01, 0x3ac0, 0x3b80, 0xfb41, 0x3900, 0xf9c1, 0xf881, 0x3840,
	0x2800, 0xe8c1, 0xe981, 0x2940, 0xeb01, 0x2bc0, 0x2a80, 0xea41,
	0xee01, 0x2ec0, 0x2f80, 0xef41, 0x2d00, 0xedc1, 0xec81, 0x2c40,
	0xe401, 0x24c0, 0x2580, 0xe541, 0x2700, 0xe7c1, 0xe681, 0x2640,
	0x2200, 0xe2c1, 0xe381, 0x2340, 0xe101, 0x21c0, 0x2080, 0xe041,
	0xa001, 0x60c0, 0x6180, 0xa141, 0x6300, 0xa3c1, 0xa281, 0x6240,
	0x6600, 0xa6c1, 0xa781, 0x6740, 0xa501, 0x65c0, 0x6480, 0xa441,
	0x6c00, 0xacc1, 0xad81, 0x6d40, 0xaf01, 0x6fc0, 0x6e80, 0xae41,
	0xaa01, 0x6ac0, 0x6b80, 0xab41, 0x6900, 0xa9c1, 0xa881, 0x6840,
	0x7800, 0xb8c1, 0xb981, 0x7940, 0xbb01, 0x7bc0, 0x7a80, 0xba41,
	0xbe01, 0x7ec0, 0x7f80, 0xbf41, 0x7d00, 0xbdc1, 0xbc81, 0x7c40,
	0xb401, 0x74c0, 0x7580, 0xb541, 0x7700, 0xb7c1, 0xb681, 0x7640,
	0x7200, 0xb2c1, 0xb381, 0x7340, 0xb101, 0x71c0, 0x7080, 0xb041,
	0x5000, 0x90c1, 0x9181, 0x5140, 0x9301, 0x53c0, 0x5280, 0x9241,
	0x9601, 0x56c0, 0x5780, 0x9741, 0x5500, 0x95c1, 0x9481, 0x5440,
	0x9c01, 0x5cc0, 0x5d80, 0x9d41, 0x5f00, 0x9fc1, 0x9e81, 0x5e40,
	0x5a00, 0x9ac1, 0x9b81, 0x5b40, 0x9901, 0x59c0, 0x5880, 0x9841,
	0x8801, 0x48c0, 0x4980, 0x8941, 0x4b00, 0x8bc1, 0x8a81, 0x4a40,
	0x4e00, 0x8ec1, 0x8f81, 0x4f40, 0x8d01, 0x4dc0, 0x4c80, 0x8c41,
	0x4400, 0x84c1, 0x8581, 0x4540, 0x8701, 0x47c0, 0x4680, 0x8641,
	0x8201, 0x42c0, 0x4380, 0x8341, 0x4100, 0x81c1, 0x8081, 0x4040
};

uint16
BlockCRC16(char *block, int16 len) 
{
	uint16 crc, i; 

	crc=CRC16_INIT;

	for (i=0; i<len; i++)
	{
		crc=(int16)((crc>>8)^(uint16)CRC16Table[(crc^(uint16)block[i])&0x00ff]);
	}

	return (crc);
}


lohs.c
/* PROGRAM:	liblohs.so
 * PURPOSE:	Specify all functions in lohs protocol
 * AUTHOR:	Igor Morgado <igor@void.com.br>
 * DATE:	2003-07-18
 * REVISED:	$Log: lohs.c,v $
 * REVISED:	Revision 1.2  2003/10/17 16:34:28  imorgado
 * REVISED:	printf changed to fprintf
 * REVISED:	
 * REVISED:	Revision 1.1.1.1  2003/10/14 18:37:21  imorgado
 * REVISED:	Initial import.
 * REVISED:	
 * Copyright 2003
 * Redistributable under the terms of the GNU General Public Licence (GPL)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <termios.h>

#include "global.h"

/* ASCII Signals */
#define 	ACK	6
#define 	NACK	21
#define		DLE	16
#define		STX	2
#define		ETX	3

/* LOHS COMMANDS */
#define		LOHS_CMD_RESET			"\x06\x00"
#define		LOHS_CMD_START_READ		"\x06\x01"
#define		LOHS_CMD_BEEP			"\x06\x03"
#define		LOHS_CMD_ENABLE_COMM_TEST	"\x06\x04"
#define		LOHS_CMD_DISABLE_COMM_TEST	"\x06\x05"
#define		LOHS_CMD_CLOCK			"\x07\x08"
#define		LOHS_CMD_SHOW_VERSION		"\x06\x0b"
#define		LOHS_CMD_BUFFER1		"\x07\x02\x01"
#define		LOHS_CMD_BUFFER2		"\x07\x02\x02"
#define		LOHS_CMD_READ_TIME1		"\x07\x06\x01"
#define		LOHS_CMD_READ_TIME2		"\x07\x06\x02"
#define		LOHS_CMD_XOR_CHECK		"\x07\x07\x00"
#define		LOHS_CMD_DISPLAY_TEXT		"\x07\x09\x40"
#define		LOHS_CMD_CRC_CHECK		"\x07\x07\x01"
#define		LOHS_CMD_XONXOFF		"\x07\x0a\x00"
#define		LOHS_CMD_RTSCTS			"\x07\x0a\x01"

/* LOHS STATUS */
#define		LOHS_STATUS_RAM_ERROR		0x10
#define		LOHS_STATUS_WAITING		0x11
#define		LOHS_STATUS_WRONG_READ		0x12
#define		LOHS_STATUS_WRONG_CLOCK		0x13
#define		LOHS_STATUS_OK			0x14
#define		LOHS_STATUS_ERROR		0x15

/************************************* 
 * 
 * External declarations 
 *
 ************************************/
extern uint16 BlockCRC16(char *block, uint16 len);


/* Global variables 
 * TODO: Expurge all global variables
 * 	and substitute everything for
 * 	professional programming
 */
struct termios oldtio;

/*************************************
 * 
 * Support functions 
 *
 * open_port: Open a serial port
 * set_serialline: Set serial line port
 * strTX: Transmits data to other peer
 * 	add CRC16 Cypher
 * strRX: Read N bytes from the peer 
 * verify_CRC16: Check CRC16 in string
 * 	based in LOHS Protocol
 *
 ************************************/

int open_port(const char *port)
{
	int fd; 
	
	fd = open (port, O_RDWR | O_NOCTTY);

	if(errno != 0)
		perror ("open port: Unable to open device"), exit(EXIT_FAILURE);
	else fcntl (fd, F_SETFL, 0);
	
	return (fd);
}

int close_port(int fd) 
{
	if(fd)
		close(fd);

	return TRUE;
}

int set_serialline(int fd,
		const char *bauds, 
		const char *flowcontrol, 
		const char *databits,
		const char *stopbits)
{
/* As defined in termios.h
 * BAUDRATE: B0, B2400, B4800, B9600, B19200, B38400, B57600
 * FLOWCONTROL: CRTSCTS (unset bit to XONXOFF)
 * DATABITS: CS7, CS8
 * STOPBITS: CSTOPB (unset bit to one stop bits)
 * PARENB: Parity enable (unset bit to no parity)
 */
	struct termios newtio;

	tcgetattr (fd, &oldtio);
        memset(&newtio, '0', sizeof(newtio));
	
	/* OLD STYLE
	newtio.c_cflag = B9600 | CRTSCTS | CS8 | CSTOPB | CLOCAL | CREAD;
	newtio.c_iflag = 0;
	newtio.c_oflag = 0;
	newtio.c_lflag = 0; */

	newtio.c_cflag = B9600 | CRTSCTS | CS8 | CSTOPB | CLOCAL | CREAD;
	newtio.c_iflag = oldtio.c_iflag;
	newtio.c_oflag = oldtio.c_oflag;
	newtio.c_lflag = oldtio.c_lflag;
	
	/* Clean the line and activate settings */
	tcsetattr(fd, TCSANOW, &newtio);
	tcflush(fd, TCIFLUSH);

	return TRUE;
}

int reset_serialline(int fd) 
{
	tcsetattr(fd, TCSANOW, &oldtio);

	return TRUE;
}

int strTX(int fd, const char *mesg)
{
	uint16 		blk;
	int 		mesg_size;
	char 		*blocked_mesg;

	mesg_size = strlen(mesg);
	
	if((blocked_mesg = (char *)malloc(mesg_size + 3)) == NULL) {
		perror("strTX: malloc()");
		return (FALSE);
	}

	strncpy(blocked_mesg, mesg, mesg_size);
	
	/* We have to add CRC16 block to the message */
	blk = BlockCRC16((char *)mesg, mesg_size);
	blocked_mesg[mesg_size]  =(byte) (blk >> 8) & 0xff;
	blocked_mesg[mesg_size+1]=(byte)  blk       & 0xff;
	blocked_mesg[mesg_size+2]='\0';
	
	if(write (fd, blocked_mesg, mesg_size + 3) < 0) {
		perror ("write() failed\n");
		free(blocked_mesg);
		return (FALSE);
	}

	free(blocked_mesg);
	return TRUE;
}

char *strRX(int fd, int nbytes)
{
	struct 	timeval tv;
	fd_set 	fds;
	char 	*mesg;
	int 	retval;
	
	if((mesg = (char *)malloc(nbytes)) == NULL) {
		perror("strRX: malloc()");
		return (NULL);
	}

	FD_ZERO(&fds);
        FD_SET(fd, &fds);
                                                                                                    
        tv.tv_sec = 5;
        tv.tv_usec = 0;
                                                                                                    
        retval = select(fd+1, &fds, NULL, NULL, &tv);
                                                                                                    
        if (retval == -1) {
                perror("strRX: select()");
		free(mesg);
		return(NULL);
	}
        else if (retval) {
		if(read (fd, mesg, nbytes) < 0) {
			perror("strRX: read()");
			free(mesg);
			return(NULL);
		}
		return mesg;
	}
        else {
		perror("strRX: read(): Timed Out\n");
		free(mesg);
		return(NULL);
        }                                                                                      
}

int verify_CRC16(const char *mesg) 
{
	uint16 	blk, mesg_size;
	char	*raw_mesg, *blocked_mesg;

	if((raw_mesg = (char *)malloc(strlen(mesg))) == NULL) {
		perror("verify_CRC16: malloc()");
		return (FALSE);
	}
	if((blocked_mesg = (char *)malloc(strlen(mesg+3))) == NULL) {
		perror("verify_CRC16: malloc()");
		return (FALSE);
	}
	
	strncpy(raw_mesg, mesg, strlen(mesg)-2);

	mesg_size=strlen(raw_mesg);
	blk=BlockCRC16(raw_mesg, mesg_size);
        blocked_mesg[mesg_size]  =(byte) (blk >> 8) & 0xff;
        blocked_mesg[mesg_size+1]=(byte)  blk       & 0xff;
        blocked_mesg[mesg_size+2]='\0';

	if (strcmp(blocked_mesg, mesg)) {
		free(blocked_mesg); free(raw_mesg);
		return TRUE;
	} else {
		free(blocked_mesg); free(raw_mesg);
		return FALSE;
	}
}

/*********************************
 * 
 * Protocol commands 
 *
 ********************************/

int LOHS_cancel_wait(int fd)
{
	strTX(fd, '\0');
	return TRUE;
}

int LOHS_set_default(int fd)
{
	LOHS_reset(fd);
	LOHS_verify_type(fd, 1); 	// CRC16 Check
	LOHS_read_times(fd, 1); 	// Single read
	LOHS_flow_control(fd, 1);	// Hardware control
	LOHS_clocks_number(fd, 0); 	// Disable clocks verify

	return TRUE;
}

int LOHS_reset(int fd)
{
	char	cmd[32], *p;

	memset(cmd, 0, sizeof(cmd));
	snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_RESET, ETX);

	if(strTX(fd, cmd) == FALSE) {
		perror("LOHS_reset: strTX");
		return (FALSE);
	}

	if((p = strRX(fd, 1)) == NULL)
		return FALSE;
	else if(atoi(p) != ACK) {
		free(p); return (FALSE);
	}

	free(p);
	sleep (1);
	return TRUE;
}

int LOHS_start_read(int fd)
{
	char	cmd[32], *p;

	memset(cmd, 0, sizeof(cmd));
	snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_START_READ, ETX);
	
	if(strTX(fd, cmd) == FALSE) {
		perror("LOHS_start_read: strTX");
		return (FALSE);
	}

	if((p = strRX(fd, 1)) == NULL) 
		return (FALSE);
	else if(atoi(p) != ACK) {
		free(p); return (FALSE);
	}
	
	free(p);
	return TRUE;
}

int LOHS_buffer_tx(int fd, int bnumber)
{
	char	cmd[32], *p;
	
	memset(cmd, 0, sizeof(cmd));
	
	switch(bnumber) {
		default:
		case 1:
			snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_BUFFER1, ETX);
			strTX(fd, cmd);
			break;
		case 2:
			snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_BUFFER2, ETX);
			strTX(fd, cmd);
			break;
	}

	if((p = strRX(fd, 1)) == NULL)
		return (FALSE);
	else if(atoi(p) != ACK ) {
		free(p); return (FALSE);
	}

	free(p);
	return TRUE;
}

int LOHS_beep(int fd)
{
	char	cmd[32], *p;
	
	memset(cmd, 0, sizeof(cmd));
	snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_BEEP, ETX);

	if(strTX(fd, cmd) == FALSE) {
		perror("LOHS_beep: strTX");
		return (FALSE);
	}
	
	if((p = strRX(fd, 1)) == NULL)
		return (FALSE);
	else if(atoi(p) != ACK) {
		free(p); return(FALSE);
	} 
	
	free(p);
	return TRUE;
}

/* 0, enable
 * 1, disable
*/
int LOHS_com_test(int fd, int value)
{
	char	cmd[32], *p;

	memset(cmd, 0, sizeof(cmd));

	switch(value) {
		default:
		case 0:
			snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_ENABLE_COMM_TEST, ETX);
			if(strTX(fd, cmd) == FALSE)
				return (FALSE);
			break;
		case 1:
			snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_DISABLE_COMM_TEST, ETX);
			if(strTX(fd, cmd) == FALSE)
				return (FALSE);
			break;
	}

	if((p = strRX(fd, 1)) == NULL)
		return (FALSE);
	else if(atoi(p) != ACK) {
		free(p); return (FALSE);
	}
	
	free(p);
	return (TRUE);
}

/* 
 * 1, single read
 * 2, double read
*/

int LOHS_read_times(int fd, int ntimes)
{
	char	cmd[32], *p;

	memset(cmd, 0, sizeof(cmd));

	switch(ntimes) {
		default:
		case 1:
			snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_READ_TIME1, ETX);
			if(strTX(fd, cmd) == FALSE)
				return (FALSE);
			break;
		case 2:
			snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_READ_TIME2, ETX);
			if(strTX(fd, cmd) == FALSE)
				return (FALSE);
			break;
	}

	if((p = strRX(fd, 1)) == NULL)
		return (FALSE);
	else if(atoi(p) != ACK) {
		free(p); return (FALSE);
	}

	free(p);
	return TRUE;
}

/*
 * 0, XOR check.
 * 1, CRC check.
*/

int LOHS_verify_type(int fd, int vtype)
{
	char	cmd[32], *p;

	memset(cmd, 0, sizeof(cmd));

	switch(vtype) {
		case 0:
			snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_XOR_CHECK, ETX);
			if(strTX(fd, cmd) == FALSE)
				return (FALSE);
			break;
		case 1:
		default:
			snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_CRC_CHECK, ETX);
			if(strTX(fd, cmd) == FALSE)
				return (FALSE);
			break;
	}
	
	if((p = strRX(fd, 1)) == NULL) 
		return (FALSE);
	else if(atoi(p) != ACK) {
		free(p); return(FALSE);
	}

	free(p);
	return TRUE;
}

int LOHS_clocks_number(int fd, int nclock)
{
	char 	cmd[32], *p;

	memset(cmd, 0, sizeof(cmd));
	
	snprintf(cmd, sizeof(cmd), "%02x%s%02x%02x", STX, LOHS_CMD_CLOCK, nclock, ETX);

	if(strTX(fd, cmd) == FALSE)
		return (FALSE);

	if((p = strRX(fd, 1)) == NULL)
		return (FALSE);
	else if(atoi(p) != ACK) {
		free(p); return (FALSE);
	}

	free(p);
	return TRUE;
}

int LOHS_read_mask(int fd, char *mask)
{
	/* Frame format 
	 * Origin: Computer
	 * <stx><08><09><MASK1><MASK2><etx><CRC1><CRC2>
	 * Response: ACK/NACK
	 * Mask 1: 7  6  5  4  3  2  1  0
	 * Mask 2: X  X  X  X 11 10  9  8
	 */
	printf ("read_mask: Not implemented\n");

	return FALSE;
}

/*
 * 0, RTS/CTS
 * 1, XON/XOFF
*/

int LOHS_flow_control(int fd, int fcontrol)
{
	char	cmd[32], *p;
	
	memset(cmd, 0, sizeof(cmd));
	
	switch(fcontrol) {
		case 0:
			snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_XONXOFF, ETX);
			if(strTX(fd, cmd) == FALSE)
				return (FALSE);
			break;
		default:
		case 1:
			snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_RTSCTS, ETX);
			if(strTX(fd, cmd) == FALSE)
				return (FALSE);
			break;
	}
		
	if((p = strRX(fd, 1)) == NULL)
		return (FALSE);
	else if(atoi(p) != ACK) {
		free(p); return (FALSE);
	}
		
	free(p);
	return TRUE;
}

int LOHS_status(int fd)
{
	/* Reader status response:
	 * Origin: Reader 
	 * Frame format:
	 * <stx><07><F0><status><etx><CRC1><CRC2>
	 * Computer response:
	 * ACK or NACK
	 *
	 * Reader status table
	 * Bit 0: Ram error
	 * Bit 1: Not Used
	 * Bit 2: Activated, but waiting card
	 * Bit 3: Wrong read (in double read)
	 * Bit 4: Wrong clocks
	 * Bit 5: Not Used
	 * Bit 6: Not Used
	 * Bit 7: Not Used
	 */ 
	char	*p;

	if((p = strRX(fd, 7)) == NULL)
		return (LOHS_STATUS_ERROR);
	
	if(verify_CRC16(p) == FALSE) {
		if(strTX(fd, (char *)NACK) == FALSE)
			return (LOHS_STATUS_ERROR);
	} else {
		if(strTX(fd, (char *)ACK) == FALSE)
			return (LOHS_STATUS_ERROR);
	}
	
	if(p[3] & 0x00) {
		free(p); return (LOHS_STATUS_RAM_ERROR);
	} else if(p[3] & 0x02) {
		free(p); return (LOHS_STATUS_WAITING);
	} else if(p[3] & 0x03) {
		free(p); return (LOHS_STATUS_WAITING);
	} else if(p[3] & 0x04) {
		free(p); return (LOHS_STATUS_WAITING);
	} else { 
		free(p); return (LOHS_STATUS_OK);
	}

	free(p);
	return (LOHS_STATUS_ERROR);
}

char * LOHS_text_block(int fd)
{
	/* Card data from reader
	 * Origin: Reader 
	 * Frame format:
	 * <stx><size><F1><text><etx><CRC1><CRC2>
	 * Computer response:
	 * ACK or NACK
	 *
	 * Text field:
	 * Each byte transports the card block, as in:
	 * 
	 * 	FIRST CARD LINE
	 * 7  6  5  4  3  2  1  0  x  x  x  11 10
	 * 	SECOND CARD LINE
	 * 9  8  7  6  5  4  3  2  1  0	
	 */ 
	 printf ("text_block: Not implemented\n");

	 return FALSE;
}

int LOHS_display_text(int fd, const char *mesg)
{
	/* Frame format:
	 * <stx><size><09><"mesg"><etx><CRC1><CRC2>
	 * Where: size== strlen(mesg)+6 
	 */
	char	cmd[32], *p;

	memset(cmd, 0, sizeof(cmd));
	snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_DISPLAY_TEXT, ETX);

	if(strTX(fd, cmd) == FALSE)
		return (FALSE);

	if((p = strRX(fd, 1)) == NULL)
		return (FALSE);
	else if(atoi(p) != ACK) {
		free(p); return (FALSE);
	}

	free(p);
	return (TRUE);
}

int LOHS_show_version(int fd)
{
	char	cmd[32], *p;

	memset(cmd, 0, sizeof(cmd));

	snprintf(cmd, sizeof(cmd), "%02x%s%02x", STX, LOHS_CMD_SHOW_VERSION, ETX);
	if(strTX(fd, cmd) == FALSE)
		return (FALSE);

	if((p = strRX(fd, 1)) == NULL) 
		return (FALSE);
	else if(atoi(p) != ACK) {
		free(p); return (FALSE);
	}
	strTX(fd, "\x02\x06\x0b\x03");

	free(p);
	return TRUE;
}

int LOHS_version_info(int fd)
{
	/* Read model, firmware and hardware version info
	 * Origin: Reader
	 * Frame format:
	 * <stx><09><F2><MODEL><FIRM><HARD><etx><CRC1><CRC2>
	 * Response:
	 * ACK or NACK
	 * Model: 	\x01 == 3m
	 * 		\x02 == 4
	 * Firmware:	\x34 == 3.4
	 * 		\x35 == 3.5
	 * Hardware:	\x00 == Standard
	 * 		\x01 == Double read support 
	 */
	unsigned char	*p;

	if((p = (unsigned char *)strRX(fd, 9)) == NULL)
		return (FALSE);

        if ( verify_CRC16(p) == FALSE ) {
		if(strTX(fd, (char *)NACK) == FALSE) {
			free(p); return (FALSE);
		}
        } else {
		if(strTX(fd, (char *)ACK) == FALSE) {
			free(p); return (FALSE);
		}
	}

	fprintf(stdout, "Model: ");
	if(p[3] == 0x01) fprintf(stdout, "3m\n");
	else if(p[3] == 0x02) fprintf(stdout, "4\n");
	else fprintf(stdout, "unknown\n");

	fprintf(stdout, "Firmware: ");
	if(p[4] == 0x34) fprintf(stdout, "3.4\n");
	else if(p[4] == 0x35) fprintf(stdout, "3.5\n");
	else fprintf(stdout, "unknown\n");

	fprintf(stdout, "Hardware: ");
	if(p[5] == 0x00) fprintf(stdout, "Standard\n");
	else if(p[5] == 0x01) fprintf(stdout, "Double Read Support\n");
	else fprintf(stdout, "unknown\n");
	
	printf("\n");
	
	free(p);
	return (TRUE);
}

